src/Core/Grid/Data/Factory/DoctrineGridDataFactory.php

```diff
@@ -93,9 +94,42 @@ public function getData(SearchCriteriaInterface $searchCriteria)
             'search_criteria' => $searchCriteria,
         ]);
 
-        $records = $searchQueryBuilder->execute()->fetchAll();
         $recordsTotal = (int) $countQueryBuilder->execute()->fetch(PDO::FETCH_COLUMN);
 
+        // If offset is out of range, we apply biggest valid offset.
+        if ($recordsTotal > 0 && $searchCriteria->getOffset() >= $recordsTotal) {
+            // Todo : those cases should be simplified with some refactoring cf. https://github.com/PrestaShop/PrestaShop/issues/29131
+
+            $searchLimit = $searchCriteria->getLimit();
+            if (empty($searchLimit) || $searchLimit > $recordsTotal) {
+                $inRangeOffset = $recordsTotal - 1;
+            } else {
+                $inRangeOffset = floor($recordsTotal / $searchLimit) * $searchLimit - 1;
+            }
+
+            // @TODO : add INFO log to prevent that we overwrite offset from `$searchCriteria->getOffset()` to `$inRangeOffset`
+
+            if ($searchCriteria instanceof Filters) {
```

Comment:
This all smells, but I get it.
I wonder if it would be an issue if we just introduce the addFilter() method in the SearchCriteriaInterface? (the method already exists in Core/Search/Filters class). :thinking:

Reply:
maybe someone could help us decide? (@jolelievre :smile: ?)

Reply:
> I wonder if it would be an issue if we just introduce the addFilter() method in the SearchCriteriaInterface? (the method already exists in Core/Search/Filters class).

I think it is a good idea. Then it would add a breaking change so we should wait until v9, shouldn't we ? :thinking: